cmd/object-api-utils.go

```diff
@@ -231,13 +232,102 @@ func pathsJoinPrefix(prefix string, elem ...string) (paths []string) {
 func pathJoin(elem ...string) string {
 	trailingSlash := ""
 	if len(elem) > 0 {
-		if HasSuffix(elem[len(elem)-1], SlashSeparator) {
+		if hasSuffixByte(elem[len(elem)-1], SlashSeparatorChar) {
 			trailingSlash = SlashSeparator
 		}
 	}
 	return path.Join(elem...) + trailingSlash
 }
 
+// pathJoinBuf - like path.Join() but retains trailing SlashSeparator of the last element.
+// Provide a string builder to reduce allocation.
+func pathJoinBuf(dst *bytes.Buffer, elem ...string) string {
+	trailingSlash := len(elem) > 0 && hasSuffixByte(elem[len(elem)-1], '/')
+	dst.Reset()
+	added := 0
+	for _, e := range elem {
+		if added > 0 || e != "" {
+			if added > 0 {
+				dst.WriteRune(SlashSeparatorChar)
+			}
+			dst.WriteString(e)
+			added += len(e)
+		}
+	}
+
+	if pathNeedsClean(dst.Bytes()) {
```

Comment:
```suggestion
	if s := pathNeedsClean(dst.Bytes()); s ! "" {
	     if trailingSlash {
	         return s + SlashSeparator
	     }
	     return s 
	}
```

is this doable? since pathNeedsClean() already peeks into the bytes.

Reply:
It is quite rare it needs cleaning after the slashes were cleaned from the input.

It was too complicated for me to figure out without too much fiddling. So I didn't want to risk breaking it.